client.go

```diff
@@ -0,0 +1,377 @@
+	req, err := c.makeRequest("POST", path, nil)
+	if err != nil {
+		return "", err
+	}
+
+	metaBytes, err := json.Marshal(&meta)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to serialize blob metadata")
+	}
+	req.Header.Add("X-Blob-Meta", base64.StdEncoding.EncodeToString(metaBytes))
+
+	redirectLocation, err := c.doWithRedirect(req)
+	if err != nil {
+		return "", err
+	}
+
+	if err := c.setMultipartRequestBody(body, req); err != nil {
+		return "", err
+	}
+
+	req.URL = redirectLocation
+	req.Header.Add("X-Blob-Meta", base64.StdEncoding.EncodeToString(metaBytes))
+
+	var response payload.PushResponse
+
+	if err := c.doJSONRequest(req, &response); err != nil {
+		return "", nil
+	}
+
+	return response.ID, nil
+}
+
+// IsHealthy returns whether the menmos cluster is healthy.
+func (c *Client) IsHealthy() (bool, error) {
+	var response payload.MessageResponse
+
+	req, err := c.makeRequest("GET", "/health", nil)
+	if err != nil {
+		return false, err
+	}
+
+	if err := c.doJSONRequest(req, &response); err != nil {
+		return false, errors.Wrap(err, "healthcheck failed")
+	}
+
+	return true, nil
+}
+
+// Query executes a query on the menmos cluster.
+func (c *Client) Query(query *payload.Query) (*payload.QueryResponse, error) {
+	var response payload.QueryResponse
+
+	request, err := c.makeJSONRequest("POST", "/query", query)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := c.doJSONRequest(request, &response); err != nil {
+		return nil, err
+	}
+
+	return &response, nil
+}
+
+// Get returns the body of the specified blob.
+// If `readRange` is non-nil, Get will attempt returns that section of the blob.
+func (c *Client) Get(blobID string, readRange *Range) (io.ReadCloser, error) {
+	if readRange != nil {
+		return &rangeReader{BlobID: blobID, Client: c, RangeStart: readRange.Start, RangeEnd: readRange.End}, nil
+	}
+
+	req, err := c.makeJSONRequest("GET", fmt.Sprintf("/blob/%s", blobID), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	redirectLocation, err := c.doWithRedirect(req)
+	if err != nil {
+		return nil, err
+	}
+
+	req.URL = redirectLocation
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	return resp.Body, nil
+}
+
+func (c *Client) GetMetadata(blobID string) (payload.BlobMeta, error) {
+	req, err := c.makeJSONRequest("GET", fmt.Sprintf("/blob/%s/metadata", blobID), nil)
+	if err != nil {
+		return payload.BlobMeta{}, err
+	}
+
+	var response payload.GetMetadataResponse
+	if err := c.doJSONRequest(req, &response); err != nil {
+		return payload.BlobMeta{}, err
+	}
+
+	if response.Metadata == nil {
+		return payload.BlobMeta{}, fmt.Errorf("get meta: blob '%s' not found", blobID)
+	}
+
+	return *response.Metadata, nil
+}
+
+// Delete deletes a blob from the cluster.
+func (c *Client) Delete(blobID string) error {
+	req, err := c.makeJSONRequest("DELETE", fmt.Sprintf("/blob/%s", blobID), nil)
+	if err != nil {
+		return err
+	}
+
+	redirectLocation, err := c.doWithRedirect(req)
```

Comment:
Sure, but how do you fix this cleanly though?

We could always add a parameter `expectRedirect bool` to `doRequest` and `doJSONRequest`, but all this blows up in our face the second we need to do something funky with the redirect (e.g.  setting the request body after the redirect, like in `pushInternal`)

or we could do a

```go
func updateRequestWithRedirectLocation(request *http.Request) error {
        redirectLocation, err := c.doWithRedirect(req)
	if err != nil {
		return err
	}

	req.URL = redirectLocation

        return req
}
```


but concretely that only save us two lines (the call to doWithRedirect and req.URL = redirectLocation) because we'd still need to check for the error when calling `redirectAndPerformRequest`, so I feel like in this case readability > deduplication.

I'm open to fixing this if you have a more elegant solution though